### ISSUE TITLE
Deprecate passing request to requirements

### ISSUE THREAD

**Issue by justanr** [justanr/flask-allows#20]
This seemed like a really good idea at the time, but the more I think about it the less I actually like this. It was a concession made for testing that is beginning to wear on me.

I'll target this at a 1.0 release and add deprecation warnings.

**Comment by jkelley05** [comment]
I was using the `request` parameter in my application and looking at `view_args` in my flask-allows function.  Since you've removed getting the `request` object as a parameter, how would you suggest I access the `request`'s `view_args` in my flask-allows functions going forward?

**Comment by justanr** [comment]
request can be safely imported and used from the flask package itself. This is how it's actually intended to be used.

So, if you had a requirement like:

    def can_access_post(user, request):
        post_id = request.view_args.get('post_id')
        # other things

You can continue using that requirement for now. However you'll get a warning when it's invoked (hopefully only the first time it's invoked, I'm nervously awaiting the big report that I've flooded someone's logs with warnings). To stop the warning, the change to make is:

    from flask import request 
    
    def can_access_post(user):
        post_id = request.view_args.get('post_id')
        # other things

Which is almost identical, with the difference being you're receiving the request object from a different location. 

The reasons for deprecating this argument are several:

1. It was a concession made strictly for testing. Typically if tests are easy to write, I think I have a good API but it turns out they lied to me this time. 🤷‍♂️

2. The request is already accessible to the requirements anyways as evidenced above. 

3. The way Allows passes the request is potentially dangerous depending on how its used in the requirements itself. It passes the context local rather than the actual request itself. If someone wrote a requirement that somehow passed that requirement into a different thread (or greenlet or similar), then that's a bug but that's on me for not warning about that.

Point 3 I didn't realize until earlier today and didn't factor into my decision, but it's honestly a better reason for ditching it. 

In any case, I'm not planning on fully removing support until the mythical 1.0 release which should give people plenty of time to migrate away from using it.

**Comment by jkelley05** [comment]
Thank you for your prompt and thoughtful response.  I expect I'll change my code base along the lines you suggested.  

As a side note: If `request` is still supposed to be passed to a flask-allows function in 0.5 you might want to check on your package on pypi.org, I was testing from a fresh build of my application and got failing tests because `request` was `None`.

**Comment by justanr** [comment]
Could you share the tests that failed? I have tests that use requirements with and without the request being passed.

**Comment by jkelley05** [comment]
Here is a rough start and I can give you more details as needed:

I import my functions to use with the flask-allows decorators with: 
`from timesheets.authorization import rules as authorizations`

The start of the route function looks like this:

```python
@offer_letters.route('/edit/<int:id_>', methods = ['GET', 'POST'])              
@login_required                                                                 
@allows.requires(authorizations.can_edit_offer_letter)                          
def edit_offer_letter(id_):   
```
If I comment out `@allows.requires(authorizations.can_edit_offer_letter)` the page will load, if I leave it in the decorator throws an exception as `request` is None

The authorizations file looks like

```python
import timesheets.authorization.checks as checks

def can_edit_offer_letter(identity, request=None, id_=None):                    
    if identity.is_anonymous:                                                   
        return False                                                            
                                                                                
    return checks.can_edit_offer_letter(identity.emplid, request = request, id_ = id_)
```

The checks function that is being called looks like:

```python
def can_edit_offer_letter(emplid, request=None, id_=None):                      
    can_edit = False                                                            
    c = OfferLetterController()                                                 
    id_ = id_ if id_ else request.view_args.get('id_')    # This is the failing line
    if is_admin(emplid):                                                        
        # Admins can always edit                                                
        can_edit = True                                                         
    else:                                                                       
        # Creators and supervisors can edit if it is not signed.                
        letter = c.get_offer_letter(id_)                                        
                                                                                
        if not letter.is_signed:                                                
            supervisors = c.get_supervisors_for_position(letter.supervisor)        
            able = [letter.created_by_emplid] + [p.emplid for p in supervisors] 
                                                                                
            if emplid in able:                                                  
                can_edit = True                                                 
                                                                                
    return can_edit                                                             
```

The exception is thrown on `request.view_args.get('id_)` and I confirmed that this code works as expected with `flask-allows==0.4`

**Comment by justanr** [comment]
```python
def can_edit_offer_letter(identity, request=None, id_=None):  
```

That'll do it, since request has a default value, [the helper I wrote to call requirements in the transitionary period](https://github.com/justanr/flask-allows/blob/96c65028de0eac677b0101e0f130bb554d38d2a4/src/flask_allows/allows.py#L129) tries calling the new style first and if it gets a type error (because it passed one argument when the receiver wants two) it calls it in the old style.

Since your requirement is written as `request=None` when the user only flow runs, Allows thinks everything worked.

My thinking on why do user-only first was:

* calling a function with the wrong number of arguments causes a TypeError
* but so do lots of other things
* however user+request is the only current option
* but since we need to catch the TypeError to toggle
* we're more likely to catch OUR TypeError if we try the user-only first.

I'm not sure of a good approach here from my perspective. If I swap the order of the calls around, I'm more likely to catch an actual TypeError from a Requirement actually blowing up rather than my artificially induced catastrophe. 

Reflection is also hard (and slow), and your function is a perfect example of that: There are three arguments and anything I write would probably detect it as an instance method (at least the first iteration would).

So I'm at an impasse. I hate that I broke your tests, but at the same time I think I'll end up breaking more things if I switch the toggle around.

:/

But I have to congratulation on you having some tests that blew up for a good reason.
